src/usb/uf2/ghostfat.c

```diff
@@ -90,50 +112,62 @@ static struct TextFile const info[] = {
     // current.uf2 must be the last element and its content must be NULL
     {.name = "CURRENT UF2", .content = NULL},
 };
-
-// code presumes each non-UF2 file content fits in single sector
-// Cannot programmatically statically assert .content length
-// for each element above.
-STATIC_ASSERT(ARRAY_SIZE(indexFile) < 512);
-
+STATIC_ASSERT(ARRAY_SIZE(infoUf2File) < BPB_SECTOR_SIZE); // GhostFAT requires files to fit in one sector
+STATIC_ASSERT(ARRAY_SIZE(indexFile)   < BPB_SECTOR_SIZE); // GhostFAT requires files to fit in one sector
 
 #define NUM_FILES          (ARRAY_SIZE(info))
 #define NUM_DIRENTRIES     (NUM_FILES + 1) // Code adds volume label as first root directory entry
+#define REQUIRED_ROOT_DIRECTORY_SECTORS ( ((NUM_DIRENTRIES+1) / DIRENTRIES_PER_SECTOR) + \
+                                         (((NUM_DIRENTRIES+1) % DIRENTRIES_PER_SECTOR) ? 1 : 0))
+STATIC_ASSERT(ROOT_DIR_SECTOR_COUNT >= REQUIRED_ROOT_DIRECTORY_SECTORS);         // FAT requirement -- Ensures BPB reserves sufficient entries for all files
+STATIC_ASSERT(NUM_DIRENTRIES < (DIRENTRIES_PER_SECTOR * ROOT_DIR_SECTOR_COUNT)); // FAT requirement -- end directory with unused entry
+STATIC_ASSERT(NUM_DIRENTRIES < BPB_ROOT_DIR_ENTRIES);                            // FAT requirement -- Ensures BPB reserves sufficient entries for all files
+STATIC_ASSERT(NUM_DIRENTRIES < DIRENTRIES_PER_SECTOR); // GhostFAT bug workaround -- else, code overflows buffer
+
+#define NUM_SECTORS_IN_DATA_REGION (BPB_TOTAL_SECTORS - BPB_RESERVED_SECTORS - (BPB_NUMBER_OF_FATS * BPB_SECTORS_PER_FAT) - ROOT_DIR_SECTOR_COUNT)
+#define CLUSTER_COUNT              (NUM_SECTORS_IN_DATA_REGION / BPB_SECTORS_PER_CLUSTER)
+
+// Ensure cluster count results in a valid FAT16 volume!
+STATIC_ASSERT( CLUSTER_COUNT >= 0x0FF5 && CLUSTER_COUNT < 0xFFF5 );
+
+// Many existing FAT implementations have small (1-16) off-by-one style errors
+// So, avoid being within 32 of those limits for even greater compatibility.
+STATIC_ASSERT( CLUSTER_COUNT >= 0x1015 && CLUSTER_COUNT < 0xFFD5 );
+
 
-#define UF2_SIZE           ((USER_FLASH_END-USER_FLASH_START) * 2)
-#define UF2_SECTORS        (UF2_SIZE / 512)
-#define UF2_FIRST_SECTOR   (NUM_FILES + 1) // WARNING -- code presumes each non-UF2 file content fits in single sector
-#define UF2_LAST_SECTOR    (UF2_FIRST_SECTOR + UF2_SECTORS - 1)
+#define UF2_FIRMWARE_BYTES_PER_SECTOR 256
+#define TRUE_USER_FLASH_SIZE (USER_FLASH_END-USER_FLASH_START)
+STATIC_ASSERT(TRUE_USER_FLASH_SIZE % UF2_FIRMWARE_BYTES_PER_SECTOR == 0); // UF2 requirement -- overall size must be integral multiple of per-sector payload?
 
-#define RESERVED_SECTORS   1
-#define ROOT_DIR_SECTORS   4
-#define SECTORS_PER_FAT    ((NUM_FAT_BLOCKS * 2 + 511) / 512)
+#define UF2_SECTORS        ( (TRUE_USER_FLASH_SIZE / UF2_FIRMWARE_BYTES_PER_SECTOR) + \
```

Comment:
the above static check already ensure it is evenly divided by UF2_FIRMWARE_BYTES_PER_SECTOR, there is no need for this computation. The 256 bytes per sector won't be changed, the flash range is only changes with 4KB page as well. We can just use it the way it is as before. Let's simplify these macro, possibly remove the `TRUE_USER_FLASH_SIZE`

Reply:
If you feel strongly, you may change it.
I would not recommend it.

I feel it is better (especially when zero code impact) to **_code defensively_**, which means to catch errors at the earliest possible time.  Every later stage an issue is found increases impact and time to fix:
* Edit  >>  Compile  >> Link  >>  Pre-execution   >>  Run-time

You have ability to edit this branch directly, so I will not stop you if you feel so strongly.

Reply:
I agree with you that we should catch error asap, however, static check right after the macro define with the same value is not catching bug IMHO. It is more on the code duplication.